Cython/Compiler/ExprNodes.py

```diff
@@ -6559,14 +6562,23 @@ def generate_result_code(self, code):
                 assert self.is_temp
                 exc_checks.append(self.type.error_condition(self.result()))
             elif func_type.exception_check != '+':
-                exc_val = func_type.exception_value
+                exc_val = func_type.exception_c_repr
                 exc_check = func_type.exception_check
                 if exc_val is not None:
                     typed_exc_val = func_type.return_type.cast_code(exc_val)
                     if func_type.return_type.is_ctuple:
                         code.globalstate.use_utility_code(UtilityCode.load_cached(
                             "IncludeStringH", "StringTools.c"))
                         exc_checks.append(f"memcmp(&{self.result()}, &{typed_exc_val}, sizeof({self.result()})) == 0")
+                    elif (self.type.is_float and (
+                            # if it isn't a Python float, it's an unknown constant
+                            not isinstance(exc_val, float) or
+                            # NaN does not compare equal to itself
+                            exc_val != exc_val)):
+                        # for floats, we may need to handle comparison with NaN
```

Comment:
I find it easier to follow the conditions with less negations.
```suggestion
                    elif self.type.is_float and not (
                            isinstance(exc_val, float) and exc_val == exc_val):
                        # Special comparison for unknown constants and NaN (because NaN != NaN).
```